cmd/backfill-redis/main.go

```diff
@@ -106,45 +111,57 @@ func main() {
-		if err != nil {
-			log.Fatalf("retrieving log uuid by index: %v", err)
-		}
-		var insertErrs []error
-		for uuid, entry := range resp.Payload {
-			// uuid is the global UUID - tree ID and entry UUID
-			e, _, _, err := unmarshalEntryImpl(entry.Body.(string))
-			if err != nil {
-				insertErrs = append(insertErrs, fmt.Errorf("error unmarshalling entry for %s: %v", uuid, err))
-				continue
-			}
-			keys, err := e.IndexKeys()
+		index := i // capture loop variable for closure
+		group.Go(func() error {
+			params := entries.NewGetLogEntryByIndexParamsWithContext(ctx)
+			params.SetLogIndex(int64(index))
+			resp, err := rekorClient.Entries.GetLogEntryByIndex(params)
 			if err != nil {
-				insertErrs = append(insertErrs, fmt.Errorf("error building index keys for %s: %v", uuid, err))
-				continue
+				log.Fatalf("retrieving log uuid by index: %v", err)
 			}
-			for _, key := range keys {
-				// remove the key-value pair from the index in case it already exists
-				if err := removeFromIndex(context.Background(), redisClient, key, uuid); err != nil {
-					insertErrs = append(insertErrs, fmt.Errorf("error removing UUID %s with key %s: %v", uuid, key, err))
+			var insertErrs []error
+			for uuid, entry := range resp.Payload {
+				// uuid is the global UUID - tree ID and entry UUID
+				e, _, _, err := unmarshalEntryImpl(entry.Body.(string))
+				if err != nil {
+					insertErrs = append(insertErrs, fmt.Errorf("error unmarshalling entry for %s: %v", uuid, err))
+					continue
 				}
-				if err := addToIndex(context.Background(), redisClient, key, uuid); err != nil {
-					insertErrs = append(insertErrs, fmt.Errorf("error inserting UUID %s with key %s: %v", uuid, key, err))
+				keys, err := e.IndexKeys()
+				if err != nil {
+					insertErrs = append(insertErrs, fmt.Errorf("error building index keys for %s: %v", uuid, err))
+					continue
+				}
+				for _, key := range keys {
+					// remove the key-value pair from the index in case it already exists
+					if err := removeFromIndex(ctx, redisClient, key, uuid); err != nil {
+						insertErrs = append(insertErrs, fmt.Errorf("error removing UUID %s with key %s: %v", uuid, key, err))
+					}
+					if err := addToIndex(ctx, redisClient, key, uuid); err != nil {
+						insertErrs = append(insertErrs, fmt.Errorf("error inserting UUID %s with key %s: %v", uuid, key, err))
+					}
+					fmt.Printf("Uploaded Redis entry %s, index %d, key %s\n", uuid, index, key)
 				}
-				fmt.Printf("Uploaded Redis entry %s, index %d, key %s\n", uuid, i, key)
 			}
-		}
-		if len(insertErrs) != 0 {
-			fmt.Printf("Errors with log index %d:\n", i)
-			for _, e := range insertErrs {
-				fmt.Println(e)
+			if len(insertErrs) != 0 {
+				fmt.Printf("Errors with log index %d:\n", index)
+				for _, e := range insertErrs {
+					fmt.Println(e)
+				}
+			} else {
+				fmt.Printf("Completed log index %d\n", index)
 			}
-		} else {
-			fmt.Printf("Completed log index %d\n", i)
-		}
+			return nil
+		})
+	}
```

Comment:
Could we also track which indices couldn't be backfilled and log those?

Reply:
Not sure if this is what you had in mind, but I added a result summary in my last commit. It does make a simple script a bit more complex, so I'm totally fine removing/changing it.

I decided to separate "parse errors" from "insert errors" when collating results, as I think parse errors are sort of unactionable and insert errors might be recoverable.

The results appear at the end of the command output, for example:
```
...
Uploaded Redis entry 24296fb24b8ad77abe5d0de220d5e64d0b52178d175a5833a5e327928368a2c9beea87009046c16d, index 21320998, key sha256:ee2ae5b13687d4d443a206359ad7c24a31b753c2233d9d1cb04dbe88535d09db
Completed log index 21320998
Backfill complete
Failed to parse 2 entries: [21320576 21320640]
Failed to insert/remove 1 entries: [21320316]
````

Reply:
This looks great!!